### ISSUE TITLE
How to reproduce the results on COCO with 4 gpus?

### ISSUE THREAD

**Issue by linhezheng19** [ultralytics/yolov5#1092]
## ❔How to reproduce the results on COCO with 4 gpus?

How can I reproduce the results on COCO with 4 gpus, any changes need to do, like the batch-size, lr, etc.? Does the lr need to change with the batch-size？Or just following the `hyp.scratch.yaml` and change the batch-size and gpu-ids?

**Comment by glenn-jocher** [comment]
@LinhzTT no changes needed to LR or anything else. You can increase your batch size to consume your available memory for fastest training. Follow multi GPU tutorial: 
https://docs.ultralytics.com/yolov5

**Comment by NanoCode012** [comment]
@LinhzTT , could you upload your reproduced results later?

**Comment by linhezheng19** [comment]
> @LinhzTT , could you upload your reproduced results later?

I re-implemented it following the original project, but it seems train slower than original project, I will upload result if i reproduce it in my implement.

**Comment by LinhzTT** [comment]
I trained from scratch on your project with `batch-size 64 2Titan XP` , it  takes about 40 mins, all 300 epochs may cost 200 hours, but 1V100 just need 2days, is this gap normal? 
And my implement version is slower, is there some tricks i need to notice? @glenn-jocher

**Comment by linhezheng19** [comment]
I trained from scratch on your project with `yolov5s.yaml batch-size 64 2Titan XP `, it takes about 40 mins, all 300 epochs may cost 200 hours, but 1V100 just need 2days, is this gap normal? @glenn-jocher 
And my implement version is slower, is there some tricks i need to notice? @glenn-jocher

**Comment by linhezheng19** [comment]
If I train yolov5 from scrach using larger batch-size like 128, any hyper-parameters need to adjust? Like use larger learning rate? @glenn-jocher

**Comment by glenn-jocher** [comment]
@LinhzTT exact code to reproduce official trained models is in README. Nothing else is required.
https://github.com/ultralytics/yolov5#training

**Comment by LinhzTT** [comment]
@glenn-jocher I see, i still have two questions.
1. I trained from scratch on your project with 'yolov5s.yaml batch-size 64  2 Titan XP' , it takes about 40 mins/epoch, all 300 epochs may cost 200 hours, but 1 V100 just need 2days, is this gap normal?
2. I re-implement version is more slower, is there some tricks speeding up that I need to notice?
Thank you for your time.

**Comment by linhezheng19** [comment]
@glenn-jocher   I see, thanks for your reply, I still have two questions.

1. I trained from scratch on your project with 'yolov5s.yaml batch-size 64 2 Titan XP' , it takes about 40 mins/epoch, all 300 epochs may cost 200 hours, but 1 V100 just need 2days, is this gap normal?
2. I re-implement version is more slower, is there some tricks speeding up that I need to notice?

Thank you for your time.

**Comment by glenn-jocher** [comment]
@LinhzTT there are no additional steps required. The commands in the README are complete. 

Speeds will vary by hardware and environment. Our reported training speeds are on a V100 on GCP. Other hardware and environments I can not comment on.

**Comment by linhezheng19** [comment]
@glenn-jocher I see, thank you very much.

**Comment by LinhzTT** [comment]
@NanoCode012 Hi! Sorry to reply you so late, I get yolov5s at mAP 35.5 in my re-implement projects with 8 TiTan XP GPUs, it is lower than original results, and I couldn't find the problem, do you have any insights?

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all        | maxDets=100 ] = 0.355
 Average Precision  (AP) @[ IoU=0.50          | area=   all        | maxDets=100 ] = 0.550
 Average Precision  (AP) @[ IoU=0.75          | area=   all        | maxDets=100 ] = 0.379
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small     | maxDets=100 ] = 0.203
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.407
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large     | maxDets=100 ] = 0.440
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all        | maxDets=  1   ] = 0.295
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all        | maxDets= 10  ] = 0.501
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all        | maxDets=100 ] = 0.558
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small     | maxDets=100 ] = 0.370
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.618
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large     | maxDets=100 ] = 0.696

**Comment by linhezheng19** [comment]
@NanoCode012 Hi! Sorry to reply you so late, I reproduced yolov5s at mAP 35.5 with **8 TiTan XP GPUs, batch-size 64, no-synbn, all hyps are same as hpt.scatch.yaml, same SGD optimizer and lr_scheduler, scale loss with num_gpus**, it is lower than original results, and I couldn't find the problem, do you have any insights?

Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.355
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.550
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.379
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.203
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.407
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.440
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.295
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.501
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.558
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.370
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.618
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.696

**Comment by glenn-jocher** [comment]
@LinhzTT models were all trained on a single GPU. We do not have official training results for multi-GPU.

Multi-GPU training should use syncbatchnorm for best results, especially at higher GPU counts.

You should not scale loss manually, all loss scaling is handled automatically.

**Comment by linhezheng19** [comment]
@glenn-jocher Scale loss is following train.py line 290.
![image](https://user-images.githubusercontent.com/20835999/100854685-4e7d4900-34c4-11eb-8d6a-4b40d5ae5769.png)

**Comment by glenn-jocher** [comment]
@LinhzTT train.py handles any loss scaling needs automatically.

**Comment by linhezheng19** [comment]
@glenn-jocher When I using 8 GPUs, I am using DDP model, and the loss while scaled with GPU counts, is the meaning of line 290 in train.py?

**Comment by glenn-jocher** [comment]
@LinhzTT train.py handles any loss scaling needs automatically.

**Comment by linhezheng19** [comment]
@glenn-jocher I use the same operation, while i using DDP model, **loss * world_size**, not setting **loss * 8** manually.

**Comment by glenn-jocher** [comment]
@LinhzTT world size is the node count, not the gpu count.

**Comment by linhezheng19** [comment]
@glenn-jocher I mistake that, generally , while we  use DDP model, we will set the node equal to gpu counts in one machine with multi-gpu, so I take them are the same in numerical.

**Comment by glenn-jocher** [comment]
@LinhzTT for multi-GPU see multi-gpu tutorial here. Nothing else is needed. No code changes are required.
https://docs.ultralytics.com/yolov5
